roofit/roofitcore/src/RooMinimizer.cxx

```diff
@@ -96,43 +97,81 @@ void RooMinimizer::cleanup()
-RooMinimizer::RooMinimizer(RooAbsReal &function, FcnMode fcnMode) : _fcnMode(fcnMode)
+/// Construcor that accepts all configuration in struct with RooAbsReal likelihood
+RooMinimizer::RooMinimizer(RooAbsReal &function, Config &cfg) : _cfg(cfg)
 {
+   // Call init functions
    initMinimizerFirstPart();
+   if (_cfg.parallelGradient || _cfg.parallelLikelihood) {
+      RooFit::TestStatistics::RooRealL *nll_real = dynamic_cast<RooFit::TestStatistics::RooRealL *>(&function);
+      if (nll_real == nullptr)
+         throw std::logic_error("In RooMinimizer constructor: Selected parallel gradient or likelihood evaluation "
+                                "but an old-style likelihood was given. Please supply NewStyle(true) as an "
+                                "argument to createNLL for new-style likelihoods.");
+      std::shared_ptr<RooFit::TestStatistics::RooAbsL> likelihood = nll_real->getRooAbsL();
+      initMultiProcess(likelihood);
+   } else {
+      initSerial(function);
+   }
+   initMinimizerFcnDependentPart(function.defaultErrorLevel());
+};
+
+/// Constructor that only accepts fcnMode configuration, for backwards compatibility
+RooMinimizer::RooMinimizer(RooAbsReal &function, FcnMode fcnMode) : _cfg()
+{
+   // Config setting
+   _cfg.fcnMode = fcnMode;
 
-   switch (_fcnMode) {
+   // Call init functions
+   initMinimizerFirstPart();
+   initSerial(function);
+   initMinimizerFcnDependentPart(function.defaultErrorLevel());
+};
+
+/// Initialize the minimizer with old style likelihoods
+void RooMinimizer::initSerial(RooAbsReal &function)
+{
+   switch (_cfg.fcnMode) {
    case FcnMode::classic: {
-      _fcn = new RooMinimizerFcn(&function, this, _verbose);
+      _fcn = std::make_unique<RooMinimizerFcn>(RooMinimizerFcn(&function, this, _cfg.verbose));
       break;
    }
    case FcnMode::gradient: {
-      _fcn = new RooGradMinimizerFcn(&function, this, _verbose);
+      _fcn = std::make_unique<RooGradMinimizerFcn>(RooGradMinimizerFcn(&function, this, _cfg.verbose));
       break;
    }
-   case FcnMode::generic_wrapper : {
+   case FcnMode::generic_wrapper: {
       throw std::logic_error("In RooMinimizer constructor: fcnMode::generic_wrapper cannot be used on a RooAbsReal! "
                              "Please use the TestStatistics::RooAbsL based constructor instead.");
    }
    default: {
       throw std::logic_error("In RooMinimizer constructor: fcnMode has an unsupported value!");
    }
    }
-
-   initMinimizerFcnDependentPart(function.defaultErrorLevel());
 }
 
-RooMinimizer::RooMinimizer(std::shared_ptr<RooFit::TestStatistics::RooAbsL> likelihood,
-                           RooFit::TestStatistics::LikelihoodMode likelihoodMode,
-                           RooFit::TestStatistics::LikelihoodGradientMode likelihoodGradientMode)
-   : _fcnMode(FcnMode::generic_wrapper)
+/// Initialize the minimizer with new style likelihoods
+void RooMinimizer::initMultiProcess(std::shared_ptr<RooFit::TestStatistics::RooAbsL> &function)
```

Comment:
```suggestion
void RooMinimizer::initMultiProcess(std::shared_ptr<RooFit::TestStatistics::RooAbsL> const& function)
```
As far as I can tell, the `function` is only used in a `const` setting here, so the method should be `const`.